File: secondary-limit-points-per-minute/python/github_endpoint_cost_audit.py
"""Compute the request rate one endpoint can sustain before it is throttled.

Read only. Every request is a GET, and the default sampled path is
/rate_limit, which does not count against the primary rate limit.

Two ceilings apply per minute to a single endpoint: 900 points, where a read
costs one point and a write costs five, and 90 seconds of CPU time per 60
seconds of real time. GitHub documents total response time as a rough estimate
of the second one, so a few timed GETs are enough to compute both and see which
binds first. That is the number a caller needs, and it is usually far lower
than 900.
"""
import argparse
import json
import logging
import os
import sys
import time

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("github_endpoint_cost_audit")

API = "https://api.github.com"
UA = "github-endpoint-cost-audit/1.0"

# Documented secondary limits for a single REST endpoint.
POINT_CAP = 900          # points per minute
CPU_CAP = 90.0           # seconds of CPU per 60 seconds of real time

# Reads cost one point; everything that changes state costs five.
CHEAP_METHODS = ("GET", "HEAD", "OPTIONS")


def points_for(method):
    """Documented point cost of one request. Pure.

    Anything unrecognised is charged the expensive rate. Guessing low here
    would produce a safe-looking ceiling for a request that is not safe, and
    the whole output of this script is a number people will pace against.
    """
    try:
        name = str(method).strip().upper()
    except (TypeError, ValueError):
        return 5
    return 1 if name in CHEAP_METHODS else 5


def cost_profile(samples):
    """Collapse timed samples into one entry per path. Pure.

    samples: [{"path", "method", "seconds"}, ...]

    Keeps the max as well as the mean because a path whose mean is comfortable
    and whose worst case is four times that will be throttled during the worst
    case and nowhere else, which is exactly the intermittent shape people
    struggle to reproduce.
    """
    grouped = {}
    for s in samples or []:
        try:
            path = str(s["path"])
            seconds = float(s["seconds"])
        except (KeyError, TypeError, ValueError):
            continue
        if seconds < 0:
            continue
        entry = grouped.setdefault(path, {"path": path, "calls": 0, "total": 0.0,
                                          "max_seconds": 0.0,
                                          "points": points_for(s.get("method", "GET"))})
        entry["calls"] += 1
        entry["total"] += seconds
        entry["max_seconds"] = max(entry["max_seconds"], seconds)

    out = {}
    for path, entry in grouped.items():
        entry["mean_seconds"] = round(entry["total"] / entry["calls"], 4)
        entry["max_seconds"] = round(entry["max_seconds"], 4)
        del entry["total"]
        out[path] = entry
    return out


def safe_rate(mean_seconds, points=1, point_cap=POINT_CAP, cpu_cap=CPU_CAP):
    """Requests per minute this endpoint sustains, and which cap binds. Pure.

    The point ceiling is a constant per method. The CPU ceiling falls as the
    endpoint gets slower, and it crosses under the point ceiling at around a
    tenth of a second a call, which is why an endpoint that feels fast can
    still be throttled at a rate nowhere near 900.
    """
    try:
        seconds = float(mean_seconds)
    except (TypeError, ValueError):
        seconds = 0.0
    points = max(1, int(points))

    by_points = point_cap / points
    by_cpu = (cpu_cap / seconds) if seconds > 0 else float("inf")

    if by_cpu < by_points:
        binding, per_minute = "cpu", by_cpu
    else:
        binding, per_minute = "points", by_points

    return {"by_points": round(by_points, 1),
            "by_cpu": None if by_cpu == float("inf") else round(by_cpu, 1),
            "binding": binding, "per_minute": round(per_minute, 1),
            "mean_seconds": round(seconds, 4), "points": points}


def verdict(path, entry, safe, configured=None):
    """Compare the computed ceiling against the rate you run at. Pure."""
    mean = safe["mean_seconds"]
    ceiling = safe["per_minute"]
    cap_name = ("the 90s-of-CPU-per-60s cap" if safe["binding"] == "cpu"
                else "the 900-points-a-minute cap")

    if configured is None:
        return ("ceiling",
                "%s costs %.3f s a call, so %s allows about %d request(s) a "
                "minute on this path." % (path, mean, cap_name, ceiling))

    try:
        configured = float(configured)
    except (TypeError, ValueError):
        return ("ceiling", "%s allows about %d a minute; no configured rate "
                           "was given to compare it against." % (path, ceiling))

    if configured > ceiling:
        return ("over-budget",
                "%s is configured for %d a minute against a ceiling of %d. %s "
                "binds first at %.3f s a call, so the surplus is refused, "
                "retried, and refused again."
                % (path, configured, ceiling, cap_name, mean))

    if configured >= ceiling * 0.8:
        return ("near-budget",
                "%s runs at %d a minute against a ceiling of %d. One slower "
                "response, or one worst case of %.3f s, closes that gap."
                % (path, configured, ceiling, entry.get("max_seconds", mean)))

    if mean >= 1.0:
        return ("expensive",
                "%s costs %.3f s a call, which caps it at %d a minute however "
                "little you are asking for today. Treat it as a path to move "
                "work off rather than a path to pace." % (path, mean, ceiling))

    return ("clear",
            "%s runs at %d a minute against a ceiling of %d, %s binding."
            % (path, configured, ceiling, cap_name))


def sample_path(session, path, count, pause):
    """Time a few sequential GETs. Sequential and paced on purpose: a sampler
    that fans out would measure the limit it is trying to describe."""
    url = API + path if path.startswith("/") else path
    samples, resource, throttled = [], None, False
    for i in range(count):
        if i:
            time.sleep(pause)
        start = time.monotonic()
        try:
            r = session.get(url, timeout=60)
        except requests.RequestException as exc:
            log.warning("%s sample %d failed: %s", path, i, exc)
            continue
        elapsed = time.monotonic() - start
        headers = {k.lower(): v for k, v in r.headers.items()}
        resource = resource or headers.get("x-ratelimit-resource")
        if r.status_code in (403, 429) and "secondary rate limit" in r.text.lower():
            throttled = True
            log.warning("%s was throttled while being measured; retry-after %s",
                        path, headers.get("retry-after", "absent"))
            continue
        if r.status_code >= 400:
            log.warning("%s sample %d returned %d", path, i, r.status_code)
            continue
        samples.append({"path": path, "method": "GET", "seconds": elapsed})
    return samples, resource, throttled


def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--path", action="append", default=None,
                    help="path to measure; repeatable (default /rate_limit)")
    ap.add_argument("--samples", type=int, default=4,
                    help="timed requests per path")
    ap.add_argument("--pause", type=float, default=1.0,
                    help="seconds between samples")
    ap.add_argument("--rate", type=float, default=None,
                    help="requests per minute your job runs at on these paths")
    args = ap.parse_args()

    token = os.environ.get("GITHUB_TOKEN")
    if not token:
        log.error("set GITHUB_TOKEN (a read-only token is enough)")
        return 2

    paths = args.path or ["/rate_limit"]
    billed = [p for p in paths if p.rstrip("/") != "/rate_limit"]
    if billed:
        log.warning("measuring %d path(s) that do cost quota: %d sample(s) "
                    "each, one point per sample", len(billed),
                    args.samples)

    session = requests.Session()
    session.headers.update({
        "Authorization": "Bearer " + token,
        "Accept": "application/vnd.github+json",
        "X-GitHub-Api-Version": "2022-11-28",
        "User-Agent": UA,
    })

    all_samples, resources, worst = [], {}, "clear"
    findings = []
    for path in paths:
        samples, resource, throttled = sample_path(session, path, max(1, args.samples),
                                                   max(0.0, args.pause))
        all_samples.extend(samples)
        if resource:
            resources[path] = resource
        if throttled:
            log.warning("%s tripped a secondary limit during measurement, which "
                        "is itself the finding: the endpoint is already over "
                        "budget at the rate this sampler used", path)

    profile = cost_profile(all_samples)
    if not profile:
        log.error("no successful samples, so there is nothing to cost")
        return 2

    ranked = sorted(profile.values(), key=lambda e: e["mean_seconds"], reverse=True)
    for entry in ranked:
        safe = safe_rate(entry["mean_seconds"], entry["points"])
        state, detail = verdict(entry["path"], entry, safe, args.rate)
        findings.append({"path": entry["path"], "state": state,
                         "mean_seconds": entry["mean_seconds"],
                         "max_seconds": entry["max_seconds"],
                         "billed_to": resources.get(entry["path"]), **safe})
        log.info("%-14s %s", state, detail)
        if resources.get(entry["path"]):
            log.info("               billed to the %s bucket",
                     resources[entry["path"]])
        if state in ("over-budget", "near-budget", "expensive"):
            worst = state if worst == "clear" else worst

    if worst != "clear":
        log.info("repair: replace per-item calls on the expensive path with one "
                 "GraphQL query returning the same fields, which is billed to a "
                 "different allowance entirely.")
        log.info("repair: raise per_page to 100 on list endpoints so the same "
                 "data arrives in a third of the calls.")
        log.info("repair: send If-None-Match with the stored etag. A 304 costs "
                 "the server almost nothing and costs you nothing at all.")
        log.info("repair: where the calls are unavoidable, spread them across "
                 "the minute rather than bursting, and on a throttled response "
                 "sleep the whole retry-after before resuming that path.")

    print(json.dumps({"findings": findings, "configured_per_minute": args.rate},
                     indent=2))
    return 1 if worst == "over-budget" else 0


if __name__ == "__main__":
    sys.exit(main())

File: secondary-limit-points-per-minute/node/github-endpoint-cost-audit.mjs
/**
 * Compute the request rate one endpoint can sustain before it is throttled.
 *
 * Read only. Every request is a GET, and the default sampled path is
 * /rate_limit, which does not count against the primary rate limit.
 *
 * Two per-minute ceilings apply to a single endpoint: 900 points, and 90
 * seconds of CPU per 60 seconds of real time. Response time is the documented
 * rough estimate of the second, so timed GETs are enough to see which binds.
 */
const API = 'https://api.github.com';
const UA = 'github-endpoint-cost-audit/1.0';

// Documented secondary limits for a single REST endpoint.
export const POINT_CAP = 900;
export const CPU_CAP = 90;

// Reads cost one point; everything that changes state costs five.
const CHEAP_METHODS = ['GET', 'HEAD', 'OPTIONS'];

/**
 * Documented point cost of one request. Pure.
 * Unrecognised methods are charged the expensive rate, because guessing low
 * produces a safe-looking ceiling for a request that is not safe.
 */
export function pointsFor(method) {
  const name = String(method ?? '').trim().toUpperCase();
  return CHEAP_METHODS.includes(name) ? 1 : 5;
}

/**
 * Collapse timed samples into one entry per path. Pure.
 * Keeps the max as well as the mean: a path with a comfortable mean and a bad
 * worst case is throttled during the worst case and nowhere else.
 */
export function costProfile(samples) {
  const grouped = new Map();
  for (const s of samples ?? []) {
    const path = s?.path === undefined ? null : String(s.path);
    const seconds = Number(s?.seconds);
    if (path === null || !Number.isFinite(seconds) || seconds < 0) continue;
    if (!grouped.has(path)) {
      grouped.set(path, {
        path, calls: 0, total: 0, max_seconds: 0, points: pointsFor(s.method ?? 'GET'),
      });
    }
    const entry = grouped.get(path);
    entry.calls += 1;
    entry.total += seconds;
    entry.max_seconds = Math.max(entry.max_seconds, seconds);
  }

  const out = {};
  for (const [path, entry] of grouped) {
    entry.mean_seconds = Math.round((entry.total / entry.calls) * 10000) / 10000;
    entry.max_seconds = Math.round(entry.max_seconds * 10000) / 10000;
    delete entry.total;
    out[path] = entry;
  }
  return out;
}

/**
 * Requests per minute this endpoint sustains, and which cap binds. Pure.
 * The CPU ceiling falls as the endpoint gets slower and crosses under the
 * point ceiling at around a tenth of a second a call.
 */
export function safeRate(meanSeconds, points = 1, pointCap = POINT_CAP, cpuCap = CPU_CAP) {
  const secs = Number.isFinite(Number(meanSeconds)) ? Number(meanSeconds) : 0;
  const p = Math.max(1, Number.parseInt(points, 10) || 1);

  const byPoints = pointCap / p;
  const byCpu = secs > 0 ? cpuCap / secs : Infinity;

  const binding = byCpu < byPoints ? 'cpu' : 'points';
  const perMinute = Math.min(byCpu, byPoints);

  return {
    by_points: Math.round(byPoints * 10) / 10,
    by_cpu: Number.isFinite(byCpu) ? Math.round(byCpu * 10) / 10 : null,
    binding,
    per_minute: Math.round(perMinute * 10) / 10,
    mean_seconds: Math.round(secs * 10000) / 10000,
    points: p,
  };
}

/** Compare the computed ceiling against the rate you run at. Pure. */
export function verdict(path, entry, safe, configured = null) {
  const mean = safe.mean_seconds;
  const ceiling = safe.per_minute;
  const capName = safe.binding === 'cpu'
    ? 'the 90s-of-CPU-per-60s cap'
    : 'the 900-points-a-minute cap';

  if (configured === null || configured === undefined) {
    return ['ceiling',
      `${path} costs ${mean.toFixed(3)} s a call, so ${capName} allows about ` +
      `${Math.trunc(ceiling)} request(s) a minute on this path.`];
  }

  const rate = Number(configured);
  if (!Number.isFinite(rate)) {
    return ['ceiling',
      `${path} allows about ${Math.trunc(ceiling)} a minute; no configured ` +
      'rate was given to compare it against.'];
  }

  if (rate > ceiling) {
    return ['over-budget',
      `${path} is configured for ${Math.trunc(rate)} a minute against a ` +
      `ceiling of ${Math.trunc(ceiling)}. ${capName} binds first at ` +
      `${mean.toFixed(3)} s a call, so the surplus is refused, retried, and ` +
      'refused again.'];
  }

  if (rate >= ceiling * 0.8) {
    return ['near-budget',
      `${path} runs at ${Math.trunc(rate)} a minute against a ceiling of ` +
      `${Math.trunc(ceiling)}. One slower response, or one worst case of ` +
      `${(entry?.max_seconds ?? mean).toFixed(3)} s, closes that gap.`];
  }

  if (mean >= 1) {
    return ['expensive',
      `${path} costs ${mean.toFixed(3)} s a call, which caps it at ` +
      `${Math.trunc(ceiling)} a minute however little you are asking for ` +
      'today. Treat it as a path to move work off rather than a path to pace.'];
  }

  return ['clear',
    `${path} runs at ${Math.trunc(rate)} a minute against a ceiling of ` +
    `${Math.trunc(ceiling)}, ${capName} binding.`];
}

const sleep = (ms) => new Promise((r) => { setTimeout(r, ms); });

/** Time a few sequential GETs. Sequential on purpose: a sampler that fanned
 * out would measure the limit it is trying to describe. */
async function samplePath(token, path, count, pause) {
  const url = path.startsWith('/') ? API + path : path;
  const samples = [];
  let resource = null;
  let throttled = false;
  for (let i = 0; i < count; i += 1) {
    if (i) await sleep(pause * 1000);
    const start = performance.now();
    let res;
    try {
      res = await fetch(url, {
        headers: {
          Authorization: `Bearer ${token}`,
          Accept: 'application/vnd.github+json',
          'X-GitHub-Api-Version': '2022-11-28',
          'User-Agent': UA,
        },
      });
    } catch (err) {
      console.warn(`${path} sample ${i} failed: ${err.message}`);
      continue;
    }
    const text = await res.text();
    const elapsed = (performance.now() - start) / 1000;
    const lowered = {};
    for (const [k, v] of res.headers.entries()) lowered[k.toLowerCase()] = v;
    resource = resource ?? lowered['x-ratelimit-resource'] ?? null;
    if ((res.status === 403 || res.status === 429)
        && text.toLowerCase().includes('secondary rate limit')) {
      throttled = true;
      console.warn(`${path} was throttled while being measured; retry-after ` +
        `${lowered['retry-after'] ?? 'absent'}`);
      continue;
    }
    if (res.status >= 400) {
      console.warn(`${path} sample ${i} returned ${res.status}`);
      continue;
    }
    samples.push({ path, method: 'GET', seconds: elapsed });
  }
  return { samples, resource, throttled };
}

async function main() {
  const token = (process.env.GITHUB_TOKEN || "dummy-github-token");
  if (!token) {
    console.error('set GITHUB_TOKEN (a read-only token is enough)');
    process.exitCode = 2;
    return;
  }
  const path = process.argv[2] ?? '/rate_limit';
  const count = Math.max(1, Number.parseInt(process.argv[3] ?? '4', 10) || 4);
  const rate = process.argv[4] === undefined ? null : Number(process.argv[4]);

  if (path.replace(/\/$/, '') !== '/rate_limit') {
    console.warn(`measuring a path that does cost quota: ${count} sample(s), ` +
      'one point each');
  }

  const { samples, resource, throttled } = await samplePath(token, path, count, 1);
  if (throttled) {
    console.warn(`${path} tripped a secondary limit during measurement, which ` +
      'is itself the finding: the endpoint is already over budget at the rate ' +
      'this sampler used');
  }

  const profile = costProfile(samples);
  const entries = Object.values(profile).sort((a, b) => b.mean_seconds - a.mean_seconds);
  if (!entries.length) {
    console.error('no successful samples, so there is nothing to cost');
    process.exitCode = 2;
    return;
  }

  const findings = [];
  let worst = 'clear';
  for (const entry of entries) {
    const safe = safeRate(entry.mean_seconds, entry.points);
    const [state, detail] = verdict(entry.path, entry, safe, rate);
    findings.push({ path: entry.path, state, max_seconds: entry.max_seconds,
      billed_to: resource, ...safe });
    console.log(`${state.padEnd(14)} ${detail}`);
    if (resource) console.log(`               billed to the ${resource} bucket`);
    if (['over-budget', 'near-budget', 'expensive'].includes(state) && worst === 'clear') {
      worst = state;
    }
  }

  if (worst !== 'clear') {
    console.log('repair: replace per-item calls on the expensive path with one ' +
      'GraphQL query, which is billed to a different allowance entirely.');
    console.log('repair: raise per_page to 100 on list endpoints so the same ' +
      'data arrives in a third of the calls.');
    console.log('repair: send If-None-Match with the stored etag. A 304 costs ' +
      'the server almost nothing and costs you nothing at all.');
    console.log('repair: where the calls are unavoidable, spread them across ' +
      'the minute rather than bursting, and sleep the whole retry-after before ' +
      'resuming that path.');
  }

  console.log(JSON.stringify({ findings, configured_per_minute: rate }, null, 2));
  process.exitCode = worst === 'over-budget' ? 1 : 0;
}

// Only run when invoked directly, so importing this from the test file does not
// start main() and set an exit code the tests never asked for.
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => { console.error(err.message); process.exitCode = 2; });
}
